scipy/linalg/src/_batched_linalg_module.cc

```diff
@@ -190,6 +197,267 @@ _linalg_solve(PyObject* Py_UNUSED(dummy), PyObject* args) {
+
+    bool dims_match = ndim_b == ndim;
+    if (dims_match) {
+        for (int i=0; i<ndim-2; i++) {
+            dims_match = dims_match && (shape[i] == shape_b[i]);
+        }
+    }
+    if (!dims_match){
+        PyErr_SetString(PyExc_ValueError, "`a` and `b` shape mismatch.");
+        return NULL;
+    }
+
+    npy_intp m = shape[ndim - 2];
+    npy_intp n = shape[ndim - 1];
+    npy_intp min_mn = m < n ? m : n;
+    npy_intp nrhs = PyArray_DIM(ap_b, ndim-1);
+
+    // Allocate the output(s)
+    npy_intp shape_1[NPY_MAXDIMS];
+    for(int i=0; i<PyArray_NDIM(ap_Am); i++) {
+        shape_1[i] = PyArray_DIM(ap_Am, i);
+    }
+
+    // x.shape = (..., N, NRHS)
+    shape_1[ndim-2] = n;
+    shape_1[ndim-1] = nrhs;
+    ap_x = (PyArrayObject *)PyArray_SimpleNew(ndim, shape_1, typenum);
+    if (!ap_x) {
+        PyErr_NoMemory();
+        return NULL;
+    }
+
+    // S array is not used by ?gelsy
+    if (strcmp(lapack_driver, "gelsy") != 0) {
+        // S.dtype is real if A.dtype is complex
+        npy_intp typenum_S = typenum;
+        if (typenum_S == NPY_COMPLEX64) { typenum_S = NPY_FLOAT32; }
+        else if (typenum_S == NPY_COMPLEX128) { typenum_S = NPY_FLOAT64; }
+
+        // S.shape = (..., min_mn)
+        shape_1[ndim - 2] = min_mn;
+        ap_S = (PyArrayObject *)PyArray_SimpleNew(ndim-1, shape_1, typenum_S);
+        if (!ap_S) {
+            PyErr_NoMemory();
+            return NULL;
+        }
+    }
+
+    // rank.shape = batch_shape
+    ap_rank = (PyArrayObject *)PyArray_SimpleNew(ndim-2, shape_1, NPY_INT64);
+    if (!ap_rank) {
+        PyErr_NoMemory();
+        return NULL;
+    }
+
+    switch(typenum) {
+        case(NPY_FLOAT32):
+            info = _lstsq<float>(ap_Am, ap_b, ap_S, ap_x, ap_rank, (float)rcond, lapack_driver, vec_status);
+            break;
+        case(NPY_FLOAT64):
+            info = _lstsq<double>(ap_Am, ap_b, ap_S, ap_x, ap_rank, rcond, lapack_driver, vec_status);
+            break;
+        case(NPY_COMPLEX64):
+            info = _lstsq<npy_complex64>(ap_Am, ap_b, ap_S, ap_x, ap_rank, (float)rcond, lapack_driver, vec_status);
+            break;
+        case(NPY_COMPLEX128):
+            info = _lstsq<npy_complex128>(ap_Am, ap_b, ap_S, ap_x, ap_rank, rcond, lapack_driver, vec_status);
+            break;
+        default:
+            PyErr_SetString(PyExc_RuntimeError, "Unknown array type.");
+            return NULL;
+    }
+
+    if (info < 0) {
+        // Some fatal error: OOM, LWORK query failed or lwork needed is too large
+        Py_DECREF(ap_x);
+        Py_DECREF(ap_rank);
+        Py_XDECREF(ap_S);
+        PyErr_SetString(PyExc_MemoryError, get_err_mesg("lstsq", lapack_driver, info).c_str());
+        return NULL;
+    }
+    PyObject *ret_lst = convert_vec_status(vec_status);
+    // with 'gelsy', we return None for `s`
+    PyObject *s_ret = ap_S != NULL ? PyArray_Return(ap_S) : Py_None;  // XXX incref None on python==3.11?
```

Comment:
I'm not 100% sure if there's a need to `Py_INCREF(Py_None)`: in python 3.12 and above `None` is immortal. 
The whole "an array or None" return value is ugly, but fixed by backwards compat, unfortunately.